File: scripts/generate-rubiks-cube.mjs
// Node.js polyfills for Three.js browser APIs
if (typeof globalThis.FileReader === "undefined") {
  globalThis.FileReader = class FileReader {
    constructor() { this.result = null; this.onload = null; }
    readAsArrayBuffer(blob) {
      blob.arrayBuffer().then((buf) => {
        this.result = buf;
        if (this.onload) this.onload({ target: this });
      });
    }
    readAsDataURL(blob) {
      blob.arrayBuffer().then((buf) => {
        const b64 = Buffer.from(buf).toString("base64");
        const type = blob.type || "application/octet-stream";
        this.result = `data:${type};base64,${b64}`;
        if (this.onload) this.onload({ target: this });
      });
    }
  };
}
if (typeof globalThis.document === "undefined") {
  globalThis.document = { createElementNS: () => ({ getContext: () => null }) };
}

import * as THREE from "three";
import { RoundedBoxGeometry } from "three/examples/jsm/geometries/RoundedBoxGeometry.js";
import { GLTFExporter } from "three/examples/jsm/exporters/GLTFExporter.js";
import { writeFileSync, mkdirSync } from "fs";

const CUBIE_SIZE = 0.95;
const CUBIE_GAP = 1.0;
const FACE_SIZE  = 0.92;
const FACE_OFFSET = CUBIE_SIZE / 2 + 0.002;
const SEGMENTS = 2;

const scene = new THREE.Scene();
scene.name = "RubiksCube_3x3";

// ── Materials ────────────────────────────────────────────
const mat = {
  interior : new THREE.MeshStandardMaterial({ color: 0x1a1a1a, roughness: 0.85, name: "interior" }),
  front    : new THREE.MeshStandardMaterial({ color: 0x009b48, roughness: 0.55, name: "green"   }),
  back     : new THREE.MeshStandardMaterial({ color: 0x0051BA, roughness: 0.55, name: "blue"    }),
  right    : new THREE.MeshStandardMaterial({ color: 0xB90000, roughness: 0.55, name: "red"     }),
  left     : new THREE.MeshStandardMaterial({ color: 0xFF5900, roughness: 0.55, name: "orange"  }),
  top      : new THREE.MeshStandardMaterial({ color: 0xFFFFFF, roughness: 0.55, name: "white"   }),
  bottom   : new THREE.MeshStandardMaterial({ color: 0xFFD500, roughness: 0.55, name: "yellow"  })
};

const faceGeo = new THREE.PlaneGeometry(FACE_SIZE, FACE_SIZE);

const pos = (i) => (i - 1) * CUBIE_GAP;

// ── Build 27 cubies ──────────────────────────────────────
for (let x = 0; x < 3; x++) {
  for (let y = 0; y < 3; y++) {
    for (let z = 0; z < 3; z++) {

      const cubie = new THREE.Group();
      cubie.name = `cubie_${x}_${y}_${z}`;
      cubie.position.set(pos(x), pos(y), pos(z));

      // Dark body
      const body = new THREE.Mesh(
        new RoundedBoxGeometry(CUBIE_SIZE, CUBIE_SIZE, CUBIE_SIZE, SEGMENTS, 0.06),
        mat.interior
      );
      body.name = "body";
      cubie.add(body);

      // Colored stickers (only on exposed faces)
      const addFace = (faceMat, px, py, pz, rx, ry) => {
        const plane = new THREE.Mesh(faceGeo, faceMat);
        plane.position.set(px, py, pz);
        if (rx) plane.rotation.x = rx;
        if (ry) plane.rotation.y = ry;
        plane.name = faceMat.name;
        cubie.add(plane);
      };

      if (z === 2) addFace(mat.front,  0, 0,  FACE_OFFSET, 0,             0);
      if (z === 0) addFace(mat.back,   0, 0, -FACE_OFFSET, 0,             Math.PI);
      if (x === 2) addFace(mat.right,  FACE_OFFSET, 0, 0,  0,             Math.PI / 2);
      if (x === 0) addFace(mat.left,  -FACE_OFFSET, 0, 0,  0,            -Math.PI / 2);
      if (y === 2) addFace(mat.top,    0, FACE_OFFSET, 0,  -Math.PI / 2,  0);
      if (y === 0) addFace(mat.bottom, 0,-FACE_OFFSET, 0,   Math.PI / 2,  0);

      scene.add(cubie);
    }
  }
}

// ── Export ────────────────────────────────────────────────
const outDir  = "models";
const outFile = `${outDir}/rubiks-cube-3x3.glb`;

mkdirSync(outDir, { recursive: true });

new GLTFExporter().parse(
  scene,
  (buffer) => {
    writeFileSync(outFile, Buffer.from(buffer));
    const kb = (Buffer.from(buffer).length / 1024).toFixed(1);
    console.log(`Generated: ${outFile}  (${kb} KB, 27 cubies)`);
  },
  (error) => {
    console.error("Export failed:", error);
    process.exit(1);
  },
  { binary: true }
);

File: scripts/generate_rubiks_cube.py
"""Generate a detachable 3x3 Rubik's Cube GLB model using trimesh."""
import os, trimesh, numpy as np

CUBIE   = 0.94   # cubie box edge
GAP     = 1.0    # center-to-center spacing
STICKER = 0.86   # sticker plane edge ( < CUBIE → visible border )
THICK   = 0.005  # sticker thickness
R       = CUBIE / 2
SR      = STICKER / 2

# ── Colours (standard Rubik's) ─────────────────────────────
C = {
    "white"  : [255, 255, 255, 255],
    "yellow" : [255, 215,   0, 255],
    "green"  : [  0, 155,  72, 255],
    "blue"   : [  0,  81, 186, 255],
    "red"    : [185,   0,   0, 255],
    "orange" : [255,  89,   0, 255],
    "dark"   : [ 18,  18,  18, 255],  # interior / border
}

def pos(i):
    return (i - 1) * GAP

def box(w, h, d, color, name=""):
    m = trimesh.creation.box(extents=[w, h, d])
    m.visual.face_colors = np.tile(color, (len(m.faces), 1))
    m.metadata["name"] = name
    return m

def sticker(axis, sign, color, name):
    """Coloured sticker plane on one face of a cubie."""
    d = [THICK, THICK, THICK]
    d[axis] = 0.0
    s = trimesh.creation.box(extents=[
        STICKER if axis != 0 else THICK,
        STICKER if axis != 1 else THICK,
        STICKER if axis != 2 else THICK,
    ])
    s.visual.face_colors = np.tile(color, (len(s.faces), 1))
    s.metadata["name"] = name

    off = R + THICK / 2
    t = np.eye(4)
    t[axis, 3] = off * sign
    s.apply_transform(t)
    return s

def cubie(x, y, z):
    """One individual cubie (body + visible stickers)."""
    parts = [box(CUBIE, CUBIE, CUBIE, C["dark"], f"cubie_{x}{y}{z}")]

    if z == 2: parts.append(sticker(2, +1, C["green"],  "F"))
    if z == 0: parts.append(sticker(2, -1, C["blue"],   "B"))
    if x == 2: parts.append(sticker(0, +1, C["red"],    "R"))
    if x == 0: parts.append(sticker(0, -1, C["orange"], "L"))
    if y == 2: parts.append(sticker(1, +1, C["white"],  "U"))
    if y == 0: parts.append(sticker(1, -1, C["yellow"], "D"))

    piece = trimesh.util.concatenate(parts)
    piece.apply_translation([pos(x), pos(y), pos(z)])
    piece.metadata["name"] = f"cubie_{x}{y}{z}"
    return piece

# ── Build scene ────────────────────────────────────────────
print("Building 27 cubies ...")
scene = trimesh.Scene()
for x in range(3):
    for y in range(3):
        for z in range(3):
            c = cubie(x, y, z)
            scene.add_geometry(c, node_name=f"cubie_{x}{y}{z}")

# ── Export GLB ─────────────────────────────────────────────
out_dir  = os.path.join(os.path.dirname(__file__), "..", "models")
os.makedirs(out_dir, exist_ok=True)
out_path = os.path.join(out_dir, "rubiks-cube-3x3.glb")

glb = scene.export(file_type="glb")
with open(out_path, "wb") as f:
    f.write(glb)

kb = os.path.getsize(out_path) / 1024
print(f"Written: {out_path}  ({kb:.1f} KB, 27 cubies)")
